Cosmos Full Text Search support

maumar opened the pull request:
- Adding model building API to configure property as full-text search enabled, as well as setup the index for it,
- Adding model validation (e.g. FTS index not matching FTS property),
- Adding EF.Functions stubs and translations for FullTextContains, FullTextContainsAll, FullTextContainsAny, FullTextScore and RRF (for hybrid),
- Adding logic in SelectExpression to produce ORDER BY RANK when necessary,
- Adding validation when attempting to mix with ORDER BY RANK with regular ORDER BY,
- Rewrite OFFSET/LIMIT from parameter to constant when ORDER BY RANK is present.
- Adding model building support for default language on the container level

Also fixed / added support for vector search on owned types (since it shares logic with FTS) and added some tests.

outstanding work:
- support for FTS Container building using Azure.ResourceManager.CosmosDb (currently blocked on updated package being released)

Fixes #35476
Fixes #35851
Fixes #35852
Fixes #35853
Fixes #35867

maumar commented on src/EFCore.Cosmos/Query/Internal/Expressions/RankOrderingExpression.cs:
```diff
@@ -0,0 +1,97 @@
+///     any release. You should only use it directly in your code with extreme caution and knowing that
+///     doing so can result in application failures when updating to a new Entity Framework Core release.
+/// </summary>
+public class RankOrderingExpression(SqlExpression expression)
```
I'm considering removing this and re-using OrderingExpression instead

maumar commented on src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs:
```diff
@@ -927,7 +927,19 @@ protected override ShapedQueryExpression TranslateCast(ShapedQueryExpression sou
         if (TranslateLambdaExpression(source, keySelector) is SqlExpression translation)
         {
-            ((SelectExpression)source.QueryExpression).ApplyOrdering(new OrderingExpression(translation, ascending));
+            if (translation is SqlFunctionExpression { IsScoringFunction: true })
```
this is a bit magical (using ORDER BY vs ORDER BY RANK based on the type of argument passed to the OrderBy linq method). Initially I had an explicit OrderByRank method, so users could be explicit about wanting to order by scoring function, but hit all sorts of issues in the nav expansion - we need process the navigations inside the scoring function and/or push includes via pending selector/orderby, but Nav expansion doesn't know about the new OrderByRank method and there is no good way to extend it :(

roji replied:
Is it possible to do `ORDER BY FullTextScore` without the RANK? If not, then I'm not sure there's a reason to want to have a completely separate OrderByRank LINQ operator... The RANK keyword can just be an expression wrapping the scoring function, or in fact be "part" of (or triggered by) the SqlFunctionExpression itself, when it's a scoring function...

maumar replied:
No, you can't do FullTextScore or RRF in a regular order by, only with RANK.

roji replied:
Right, so that really does mean that RANK has no "meaning" - it's just a thing you have to put once you use a scoring function. So then there's no real reason or need to expose an OrderByRank to the user (it's not like they have any choice here, or would be able to express different things with OrderBy vs. OrderByRank).

maumar commented on src/EFCore.Cosmos/Query/Internal/Expressions/SqlFunctionExpression.cs:
```diff
@@ -24,10 +24,21 @@ public SqlFunctionExpression(
+    {
+    }
+
+    internal SqlFunctionExpression(
```
will add to SqlExpressionFactory later

maumar commented on src/EFCore.Cosmos/Storage/Internal/CosmosClientWrapper.cs:
```diff
@@ -276,9 +298,26 @@ private static async Task<bool> CreateContainerIfNotExistsOnceAsync(
         }
 
-        if (vectorIndexes.Any())
+        if (vectorIndexes.Any()/* || fullTextIndexes.Any()*/)
```
will figure out the best way here (i.e. together or separate) - need a few more hybrid search tests

maumar commented on test/EFCore.Cosmos.FunctionalTests/FullTextSearchCosmosTest.cs:
```diff
@@ -0,0 +1,592 @@
+            message);
+    }
+
+    [ConditionalFact(Skip = "issue #35867")]
```
In some of my tests vs live service, cosmos encouraged me to use Skip/Take in order to prevent excessive use of resources due to non-streaming nature of those queries. Also, all the samples using ORDER BY RANK also use TAKE, so the fact that it currently doesn't work might be a serious problem.

maumar commented on src/EFCore.Cosmos/Storage/Internal/CosmosClientWrapper.cs:
```diff
@@ -276,9 +298,26 @@ private static async Task<bool> CreateContainerIfNotExistsOnceAsync(
         }
 
-        if (vectorIndexes.Any())
+        if (vectorIndexes.Any()/* || fullTextIndexes.Any()*/)
```
good suggestion but already on my radar

SamMonoRT commented on src/EFCore.Cosmos/Extensions/CosmosDbFunctionsExtensions.cs:
```diff
@@ -52,6 +52,41 @@ public static T CoalesceUndefined<T>(
 
+    /// <summary>
+    /// TODO
+    /// </summary>
```
leverage CoPilot to fill these in, and then review quickly

maumar commented on src/EFCore.Cosmos/Query/Internal/Expressions/RankOrderingExpression.cs:
```diff
@@ -0,0 +1,97 @@
+///     any release. You should only use it directly in your code with extreme caution and knowing that
+///     doing so can result in application failures when updating to a new Entity Framework Core release.
+/// </summary>
+public class RankOrderingExpression(SqlExpression expression)
```
We could use OrderingExpression and set IsAscending to true, cause we don't generate ASC for it, just DESC for descending. But it feels a bit hacky. Upside is less public surface. Thoughts?

roji replied:
Not sure I understand - doesn't OrderingExpression with IsAscending=true already mean something else?

In any case, IIUC we need to actually output an additional RANK keyword in SQL (so `ORDER BY RANK rrf(...)`), so it makes sense to have an expression representing that (if we go down this route I'd shorten and call this simply RankExpression).

But a lighter alternative would be to simply fold the RANK word into the SqlFunctionExpression representing the scoring function; IIUC you already have an IsScoringFunction on it, so the distinction is there, and I'm not sure you also need a special expression to wrap it?

One additional thing I'm not sure about, is why SelectExpression needs to have an additional special "clause" (RankOrdering) just to hold this - is there any special reason the scoring function can't just go into the regular Ordering clause? After all, it really is just part of the ORDER BY - we just need to generate an extra RANK before it in SQL generation, plus add some checks (no descending, only one ordering etc.).

maumar replied:
Folding RANK into the function (as I understand it: generate RANK keyword when processing SqlFunctionExpression with IsScoringFunction = true in the sql generator) is problematic for the RRF case, when you have RRF(FTScore(x.Name1), FTScore(x.Name2)), all 3 functions are scoring functions but we only need one RANK. Instead, we could make the decision when processing OrderBy in the ComosSqlGenerator - peek into the orderings and if there is only one ordering and it's a scoring function, add RANK keyword before processing the ordering itself. I can definitely try that - it reduces public surface, so that's always nice.

I added the new clause (rank ordering) because the existing clause takes multiple expressions (rank only supports one), and forces us to define ascending/descending, which rank doesn't support (and don't plan to in the future). That's where my initial "hacky" comment is coming from. It's not a perfect match, but we should be able to make it work

roji replied:
Yep, makes sense - I'm not sure anything is more is needed than a simple check in SQL generation for a SelectExpression with a single score function ordering, and then generate RANK for that case before falling through to the generic ordering logic. Pattern matching should make that quite easy to match wherever you need to. We'd simply never have a scoring function inside an ordering that contains an additional thing, nor DESC (validation would take care of that).

roji commented on src/EFCore.Cosmos/EFCore.Cosmos.csproj:
```diff
@@ -12,6 +12,7 @@
     <NoWarn>$(NoWarn);EF9101</NoWarn> <!-- Metrics is experimental -->
     <NoWarn>$(NoWarn);EF9102</NoWarn> <!-- Paging is experimental -->
     <NoWarn>$(NoWarn);EF9103</NoWarn> <!-- Vector search is experimental -->
```
Yeah, you're right - I think it's no longer experimental. Open an issue for us to make this non-experimental for 10?

maumar replied:
https://github.com/dotnet/efcore/issues/35886

roji commented:
@maumar just to note that I'm holding back from reviewing since this is still in draft (you seemed to maybe want people to take a look in yesterday's meeting). Let me know if you want a review etc.

maumar commented:
@roji feel free to hold off on reviewing a draft. I'm pretty close to marking it for review I think

roji commented on src/EFCore.Cosmos/Extensions/CosmosIndexBuilderExtensions.cs:
```diff
@@ -96,4 +96,84 @@ public static bool CanSetVectorIndexType(
+    /// <param name="indexBuilder">The builder for the index being configured.</param>
+    /// <param name="value">The value indicating whether the index is configured for Full-text search.</param>
+    /// <returns>A builder to further configure the index.</returns>
+    public static IndexBuilder ForFullText(this IndexBuilder indexBuilder, bool? value = true)
```
We have IsClustered, IsCreatedOnline etc. for SQL Server indexes, shouldn't this be called IsFullText? /cc @AndriySvyryd for naming consistency power.

AndriySvyryd replied:
Doesn't make a lot of sense grammatically, but it is consistent

maumar replied:
I was basing the naming on the vector work, which has ForVectors

roji replied:
OK. We can decide to change this as it's all still experimental (@AndriySvyryd seems to agree above...).

roji replied:
IsVectors definitely doesn't work though. I'd just go with IsFullTextIndex and IsVectorIndex...

maumar replied:
I like IsFullTextIndex and IsVectorIndex!

roji replied:
@AndriySvyryd?

maumar replied:
will fix FTS here, filed https://github.com/dotnet/efcore/issues/35895 for the vector

roji commented on src/EFCore.Cosmos/Extensions/CosmosPropertyExtensions.cs:
```diff
@@ -128,4 +128,50 @@ public static void SetVectorType(this IMutableProperty property, CosmosVectorTyp
+    /// <param name="property">The property.</param>
+    /// <returns>Returns the definition of the vector stored in this property.</returns>
+    [Experimental(EFDiagnostics.CosmosFullTextSearchExperimental)]
+    public static string? GetFullTextSearchLanguage(this IReadOnlyProperty property)
```
Does this settting belong on the property, or more on the index?

maumar replied:
per cosmos docs (https://learn.microsoft.com/en-us/azure/cosmos-db/gen-ai/full-text-search#full-text-policy), language is defined in container properties, rather than indexing policy

roji replied:
It would still be good to actually understand what the property/index split means exactly, and why the language is set on the property rather than on the index. For example, does it make sense for a user to define a "full text search language" on a property without also defining an FTS index on that same property? Or is the intention here to allow users to define multiple FTS indexes on the same property (with different indexing settings or something?), but to force them to have the same language?

Maybe reach out to the Cosmos people to understand this better? (you can loop me in)

roji commented on src/EFCore.Cosmos/Extensions/CosmosIndexExtensions.cs:
```diff
@@ -61,4 +61,49 @@ public static void SetVectorIndexType(this IMutableIndex index, VectorIndexType?
+    /// </summary>
+    /// <param name="index">The index.</param>
+    /// <returns>The index type to use, or <see langword="null" /> if none is set.</returns>
+    public static bool? IsFullTextIndex(this IReadOnlyIndex index)
```
Missing [Experimental] here and elsewhere

roji commented on src/EFCore.Cosmos/Infrastructure/Internal/CosmosModelValidator.cs:
```diff
@@ -570,6 +570,24 @@ protected virtual void ValidateIndexes(
                                 index.Properties[0].Name));
                     }
                 }
+                else if (index.FindAnnotation(CosmosAnnotationNames.FullTextIndex) != null)
```
Should we be using the extension methods rather than accessing the annotations directly? /cc @AndriySvyryd

maumar replied:
fixed for FTS, filed https://github.com/dotnet/efcore/issues/35897 to cleanup the rest

roji commented on src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs:
```diff
@@ -628,7 +628,7 @@ protected override ShapedQueryExpression TranslateCast(ShapedQueryExpression sou
         // fine. Since subqueries must be correlated (i.e. reference an array in the outer query), we use that to decide whether to
         // warn or not.
-        if (select.Orderings.Count == 0 && !_subquery)
+        if (select.Orderings.Count == 0 && !_subquery && select.RankOrdering == null)
```
If you put scoring functions inside the regular Ordering instead of a special new RankOrdering (see comment above), the need to change this goes away.

roji commented on src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.InExpressionValuesExpandingExpressionVisitor.cs:
```diff
@@ -53,4 +53,36 @@ protected override Expression VisitExtension(Expression expression)
         }
     }
+
+    private sealed class HybridSearchParameterRewriter(
```
We can just merge these two visitors into a single one to avoid doing two passes on the tree needlessly (ParameterInliner?)

roji commented on src/EFCore.Cosmos/Query/Internal/Expressions/SelectExpression.cs:
```diff
@@ -393,6 +417,12 @@ public void ApplyOrdering(OrderingExpression orderingExpression)
     /// </summary>
     public void AppendOrdering(OrderingExpression orderingExpression)
     {
+        if (_rankOrdering != null)
```
Note that you have other error checking happening in CosmosQueryableMethodTranslatingEV - we should decide whether to put it all the validation there or here (I don't think it matters, but a single place would be better...)

roji commented on src/EFCore.Cosmos/Query/Internal/Translators/CosmosFullTextSearchTranslator.cs:
```diff
@@ -0,0 +1,130 @@
+
+    private static readonly Dictionary<MethodInfo, string> SupportedMethodsMap = new()
+    {
+        { _fullTextContainsMethod, nameof(CosmosDbFunctionsExtensions.FullTextContains) },
```
Avoid nameof in this kind of situation - if we ever rename the EF .NET method names, we definitely do **not** want to generate different SQL: these are fixed, external names that come from the Cosmos documentation.

roji commented on src/EFCore.Cosmos/Query/Internal/Translators/CosmosFullTextSearchTranslator.cs:
```diff
@@ -0,0 +1,130 @@
+        if (method == _fullTextScoreMethod)
+        {
+            return new SqlFunctionExpression(
+                SupportedMethodsMap[method],
```
Does the dictionary provide any specific value here over just sticking the name of the Cosmos function in-line here?

maumar replied:
removed

roji commented on src/EFCore.Cosmos/Query/Internal/Translators/CosmosFullTextSearchTranslator.cs:
```diff
@@ -0,0 +1,130 @@
+        IReadOnlyList<SqlExpression> arguments,
+        IDiagnosticsLogger<DbLoggerCategory.Query> logger)
+    {
+        if (!SupportedMethodsMap.ContainsKey(method))
```
Not sure this check makes sense given that you're simply comparing the methods one by one... I'd consider a simple check that the DeclaringType on the MethodInfo is CosmosDbFunctionsExtensions, where all these methods are.

roji reviewed (changes requested):
Ended up doing a quick review even though the PR is still in draft... Here are some comments.

maumar commented on src/EFCore.Cosmos/Storage/Internal/CosmosDatabaseCreator.cs:
```diff
@@ -154,7 +149,36 @@ private static IEnumerable<ContainerProperties> GetContainersToCreate(IModel mod
+                }
+            }
+
+            foreach (var ownedType in entityType.GetNavigations().Where(x => x.ForeignKey.IsOwnership && !x.IsOnDependent).Select(x => x.TargetEntityType))
```
is there a better way to do this? @AndriySvyryd

AndriySvyryd replied:
No, this is as good as it gets. Also check that the target entity type is not a root document, just in case.

roji commented on src/EFCore.Cosmos/Query/Internal/Expressions/SelectExpression.cs:
```diff
@@ -393,6 +399,22 @@ public void ApplyOrdering(OrderingExpression orderingExpression)
     {
+        var existingScoringFunctionOrdering = _orderings.Any(x => x.Expression is SqlFunctionExpression { IsScoringFunction: true });
+
+        if (orderingExpression.Expression is SqlFunctionExpression { IsScoringFunction: true }
```
This seems to assume that AppendOrdering will only ever get called if there's already something in `_orderings` (because it throws immediately if the argument is a scoring function); is that true, do we want to make it? In other words, I'd make this not throw for the case that `_orderings` is empty.

maumar replied:
there *should* always be something in the orderings - we only call it from TranslateThenBy, but makes sense to make it more defensive in case things change in the future

roji commented on src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.InExpressionValuesExpandingExpressionVisitor.cs:
```diff
@@ -49,6 +49,26 @@ protected override Expression VisitExtension(Expression expression)
                     ? sqlExpressionFactory.ApplyDefaultTypeMapping(sqlExpressionFactory.Constant(false))
                     : sqlExpressionFactory.In((SqlExpression)Visit(inExpression.Item), values);
             }
+            else if (expression is SelectExpression { Orderings: [{ Expression: SqlFunctionExpression { IsScoringFunction: true } }] } hybridSearchSelect
```
nit: can be a switch with InExpression above

roji commented on src/EFCore.Cosmos/Query/Internal/CosmosQuerySqlGenerator.cs:
```diff
@@ -341,6 +341,11 @@ protected override Expression VisitSelect(SelectExpression selectExpression)
         {
             _sqlBuilder.AppendLine().Append("ORDER BY ");
 
+            if (selectExpression.Orderings is [{ Expression: SqlFunctionExpression { IsScoringFunction: true } }])
```
We can include a debug assertion that if this condition isn't true, there are no scoring functions anywhere (just to make the situation clearer to the reader, and in case we have a bug earlier)

roji commented on src/EFCore.Cosmos/Query/Internal/SqlExpressionFactory.cs:
```diff
@@ -660,6 +660,28 @@ public virtual SqlExpression Function(
+    ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+    /// </summary>
+    [Experimental(EFDiagnostics.CosmosFullTextSearchExperimental)]
+    public virtual SqlExpression ScoringFunction(
```
Unrelated to this PR: I still don't understand what it gets us to instantiate via SqlExpressionFactory, at least in this case; seems a bit like a cargo cult we fell into.

maumar replied:
don't need to worry about type mappings as SqlExpressionFactory handles those, but otherwise yeah - John Frum told me to use it.

roji commented on src/EFCore.Cosmos/Query/Internal/Translators/CosmosFullTextSearchTranslator.cs:
```diff
@@ -0,0 +1,73 @@
+        return method.Name switch
+        {
+            nameof(CosmosDbFunctionsExtensions.FullTextContains)
+                when arguments is [_, _, _] => sqlExpressionFactory.Function(
```
Might as well take the arguments as part of the pattern matching and pass them below, instead of referencing `arguments[1]` again etc. (like you do below with Rrf). This makes it super clear what parameters the function actually accepts etc.

```suggestion
                when arguments is [_, var property, var keywords] => sqlExpressionFactory.Function(
```

roji commented on src/EFCore.Cosmos/Query/Internal/Translators/CosmosFullTextSearchTranslator.cs:
```diff
@@ -0,0 +1,73 @@
+                    typeMappingSource.FindMapping(typeof(double))),
+
+            nameof(CosmosDbFunctionsExtensions.FullTextContainsAny) or nameof(CosmosDbFunctionsExtensions.FullTextContainsAll)
+                when arguments is [_, SqlExpression containsAllAnyPath, ArrayConstantExpression containsAllAnyArguments] => sqlExpressionFactory.Function(
```
I see, Cosmos doesn't seem to support parameterizing the list of keywords for FullTextContainsAny/All (IIUC individual keywords can be parameterized, but the list itself cannot)...

I'd at least add a check and throw an informative exception (currently this will throw "could not translate"). Though it should be easy to make this work, i.e. stick a parameterized array expression here, and inline it later (in the parameter inliner). Thoughts?

nit: can call the variable `keywords` instead of `containsAllAnyArguments`, just like the FullTextContainsAny parameter name.

maumar replied:
yeah, individual keywords can be parameters, we have test for that. WRT parameterizing the entire keywords list, you are right, it was pretty easy to make it work, although I only made it work for FullTextContainsAll/Any (matching those specific function names). It seems a little smelly because ParameterInliner now knows about specific sql functions, which it probably shouldn't care.

roji replied:
> It seems a little smelly because ParameterInliner now knows about specific sql functions, which it probably shouldn't care.

FWIW we have the same in the relational nullability processor IIRC, with SUM :/

Yeah, it's maybe not ideal, but this is alll the internal workings of a single provider - I think it's OK.

roji reviewed (changes requested):
/cc @AndriySvyryd in case you want to take a quick look for the metadata side and Cosmos container building part.

maumar commented on src/EFCore.Cosmos/Query/Internal/Expressions/SelectExpression.cs:
```diff
@@ -393,6 +399,22 @@ public void ApplyOrdering(OrderingExpression orderingExpression)
     /// </summary>
     public void AppendOrdering(OrderingExpression orderingExpression)
     {
+        var existingScoringFunctionOrdering = _orderings.Any(x => x.Expression is SqlFunctionExpression { IsScoringFunction: true });
```
> Also, this looks like it can move inside the if

you mean this one? `if (_orderings.FirstOrDefault(o => o.Expression.Equals(orderingExpression.Expression)) == null)`

maumar replied:
we are also using it to decide which exception to throw (either multiple scoring functions so we say to use RRF, or scoring function with another order by, in which case we say u can't mix those).

AndriySvyryd commented on src/EFCore.Cosmos/Storage/Internal/CosmosClientWrapper.cs:
```diff
@@ -290,6 +322,23 @@ private static async Task<bool> CreateContainerIfNotExistsOnceAsync(
+        {
+            var ownership = entityType.FindOwnership()!;
+
+            return GetPathFromRoot(ownership.PrincipalEntityType) + "/" + ownership.GetNavigation(pointsToPrincipal: false)!.Name;
```
Use JSON property name.
Also, for collections you need to add `/[]`

maumar replied:
wildcards are not supported for FTS/vector paths, I disabled the scenario instead

AndriySvyryd commented on test/EFCore.Cosmos.FunctionalTests/HybridSearchCosmosTest.cs:
```diff
@@ -0,0 +1,215 @@
+                    bb.Property(x => x.AnotherDescription).IsFullText();
+                    bb.HasIndex(x => x.AnotherDescription).ForFullText();
+                });
+            });
```
Add tests for owned collections

maumar commented:
@roji @AndriySvyryd the PR is now ready for review.

AndriySvyryd commented on src/EFCore.Cosmos/Extensions/CosmosPropertyExtensions.cs:
```diff
@@ -128,4 +128,50 @@ public static void SetVectorType(this IMutableProperty property, CosmosVectorTyp
+    /// <param name="language">The full-text search language for the property.</param>
+    [Experimental(EFDiagnostics.CosmosFullTextSearchExperimental)]
+    public static void SetFullTextSearchLanguage(this IMutableProperty property, string? language)
+        => property.SetOrRemoveAnnotation(CosmosAnnotationNames.FullTextSearchLanguage, language);
```
```suggestion
        => property.SetAnnotation(CosmosAnnotationNames.FullTextSearchLanguage, language);
```
Same below

roji commented on src/EFCore.Cosmos/Query/Internal/Expressions/SelectExpression.cs:
```diff
@@ -393,6 +400,18 @@ public void ApplyOrdering(OrderingExpression orderingExpression)
+                || existingScoringFunctionOrdering)
+            {
+                throw new InvalidOperationException(existingScoringFunctionOrdering
+                    ? CosmosStrings.OrderByMultipleScoringFunctionWithoutRrf(nameof(CosmosDbFunctionsExtensions.Rrf))
```
Looks like this throws the wrong message if there's an existing scoring function and the user tries to append a non-scoring function. We should only tell them to use RRF if both the existing and the new ordering are scoring functions, otherwise they're just mixing scoring with non-scoring.

maumar replied:
I even have the test for this scenario and was validating against wrong message 🤦‍♂️

roji commented on src/EFCore.Cosmos/Query/Internal/Translators/CosmosFullTextSearchTranslator.cs:
```diff
@@ -0,0 +1,81 @@
+            return null;
+        }
+
+        return method.Name switch
```
Looks nice no? I really like what switch expressions and pattern matching have done to the language...

maumar replied:
yeah, especially the collection stuff. Still struggle to read the code sometimes, but it sure is easier to write :D

roji reviewed (changes requested):
@maumar see batch of comments.

I think we should maybe discuss the FTS property vs. index question in design...

maumar commented on src/EFCore.Cosmos/Extensions/CosmosIndexBuilderExtensions.cs:
```diff
@@ -28,6 +27,7 @@ public static class CosmosIndexBuilderExtensions
     /// <param name="indexBuilder">The builder for the index being configured.</param>
     /// <param name="indexType">The type of vector index to create.</param>
     /// <returns>A builder to further configure the index.</returns>
+    [Experimental(EFDiagnostics.CosmosVectorSearchExperimental)]
```
the whole CosmosIndexBuilderExtensions was vector experimental. I broke it down to vector + FTS in the methods, and will remove the vector part in future pr

maumar commented on src/EFCore.Cosmos/Properties/CosmosStrings.Designer.cs:
```diff
@@ -95,6 +103,14 @@ public static string ContainerNotOnRoot(object? entityType, object? baseEntityTy
             => GetString("CosmosNotInUse");
 
+        /// <summary>
+        ///     Creating a container with full-text search or vector properties inside a collection navigation is currently not supported on EF Core. Path: '{path}'.
```
It's kinda supported in Cosmos - wildcards when defining paths are not supported so you must define index individually for each element in the collection:

/Entity/Collection/0/FTSProp
/Entity/Collection/1/FTSProp
/Entity/Collection/2/FTSProp
/Entity/Collection/3/FTSProp

rather than:
/Entity/Collection/[]/FTSProp

roji replied:
OK, understood - thanks.

maumar replied:
Added a statement suggesting using Cosmos SDK to build the container. Will also make sure to talk about this case in the docs

maumar commented on src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.InExpressionValuesExpandingExpressionVisitor.cs:
```diff
@@ -9,48 +9,94 @@ namespace Microsoft.EntityFrameworkCore.Cosmos.Query.Internal;
+                case SqlFunctionExpression
+                {
+                    Name: "FullTextContainsAny" or "FullTextContainsAll",
+                    Arguments: [var property, SqlParameterExpression { TypeMapping: { ElementTypeMapping: not null } } keywords]
```
true. I do use element type mapping inside the method though, so instead of null check I will assign to variable

maumar commented on src/EFCore.Cosmos/Extensions/CosmosPropertyBuilderExtensions.cs:
```diff
@@ -243,4 +243,95 @@ private static CosmosVectorType CreateVectorType(DistanceFunction distanceFuncti
+    /// <param name="language">The language for the full-text search.</param>
+    /// <returns>The same builder instance so that multiple calls can be chained.</returns>
+    [Experimental(EFDiagnostics.CosmosFullTextSearchExperimental)]
+    public static PropertyBuilder IsFullText(
```
What I don't like about HasFullTextSearchLanguage is that users never actually set the language currently. Might make more sense once they support more languages, so there's actually a choice.

roji replied:
If Cosmos supports setting the language but only support English for now, then we'd just be exposing Cosmos on this. I personally don't see an issue if that's the case - once Cosmos introduces other languages, things would already be ready and we wouldn't have to do yet another feature (which may take months to get released in the next major version...).

Otherwise, why introduce this at all rather than introducing something that's inaccurately named? Can/should we just leave it out for now?

maumar commented:
@roji @AndriySvyryd I updated the PR addressing the Property/Index API and fixing a few small errors. You can review just the second commit - all the changes are there. 

- I put the HasDefaultFullTextLanguage setting on the entity type builder, rather than creating a container builder just for it. We already have some container-level settings (e.g. TTL) and we defined them on entity type builder, so going for consistency here. 

- Added tests to make sure values are propagated correctly (if property specifies language, use that, otherwise use default, otherwise use en-US). These tests need to be adjusted once cosmos supports more languages, for now they all just throw Cosmos exception complaining about not supported language. I manually verified that we wire-up things correctly when creating the container.

- Added test for using full text property without the index and removed validation preventing this scenario.

- Conditionally skipping the tests when running on live service due to Azure.ResourceManager.CosmosDb not supporting FTS.

maumar commented on src/EFCore.Cosmos/Infrastructure/Internal/CosmosModelValidator.cs:
```diff
@@ -330,6 +331,27 @@ protected virtual void ValidateSharedContainerCompatibility(
                 }
             }
+
+            var currentFullTextSearchDefaultLanguage = entityType.GetDefaultFullTextSearchLanguage();
```
I added validation for this here, rather than during container creation (where all other FTS model validation resides). This seems like a proper model configuration issue, rather than something that will change (i.e. will be supported in the future) on Cosmos side. Thoughts? @AndriySvyryd

roji commented on src/EFCore.Cosmos/Extensions/CosmosEntityTypeExtensions.cs:
```diff
@@ -586,4 +587,52 @@ public static void SetThroughput(this IMutableEntityType entityType, int? throug
+    /// <param name="entityType">The entity type.</param>
+    /// <returns>The default language for the full-text search.</returns>
+    [Experimental(EFDiagnostics.CosmosFullTextSearchExperimental)]
+    public static string? GetDefaultFullTextSearchLanguage(this IReadOnlyEntityType entityType)
```
We could even have a default language in the model if we wanted to - but not critical for this PR obviously.

maumar replied:
https://github.com/dotnet/efcore/issues/35929

roji commented on src/EFCore.Cosmos/Extensions/CosmosPropertyBuilderExtensions.cs:
```diff
@@ -243,4 +243,95 @@ private static CosmosVectorType CreateVectorType(DistanceFunction distanceFuncti
+    /// <param name="language">The language for the full-text search.</param>
+    /// <returns>The same builder instance so that multiple calls can be chained.</returns>
+    [Experimental(EFDiagnostics.CosmosFullTextSearchExperimental)]
+    public static PropertyBuilder IsFullTextProperty(
```
I still find the discrepancy between builder (IsFullTextProperty) and metadata (SetFullTextSearchLanguage) names odd and unclear. For example, what does IsFullTextProperty() actually do when a language isn't passed? Since we just call SetFullTextSearchLanguage with null, it doesn't seem to actually do anything (apart erasing a previously-set language); if so, then this method doesn't actually make the property into a full-text property - it really just sets the language.

Here's another way to look at it. Since Cosmos allows defining a default language at the container level, I'm assuming it's possible to omit it from the property level (otherwise why have a default at all). If that's the case (should confirm), then there needs to be a way to way "this is a full text property" without setting a language.

If the above is true on the Cosmos side - and it needs to be confirmed by simple experimentation - we simply need to have two separate metadata: one "IsFullTextSearchEnabled" (boolean), and one "FullTextSearchLanguage" (which is optional, if not specified the container default applies). We could still of course have a single convenience builder method (I'd call it EnableFullTextSearch), which sets **both** metadata. According to my understand of how Cosmos actually works - that the language specification on the property level is optional - this is what we need.

@maumar does this correspond to the Cosmos support as you understand it? @AndriySvyryd does this make sense?

---

On another note, an alternative possible naming for the builder API could simply be EnableFullTextSearch(), since the property still remains a regular property that can be used for other things (xmldoc can also be updated to reflect that). Constrast this with vector properties, which have no other possible use.

maumar replied:
> One last possibility here is to use a null language to mean "this is a full text property who's language hasn't been set by the user (and should default to the container default" (is this the current design @maumar?). If so, then there's no way to turn off a property being a full-text property (since passing a null language activate FTS).

you are right, this is the current design - will switch to 2 annotations, it is much cleaner

AndriySvyryd replied:
If I understood the design correctly there is no such setting as "this is an FTS property", FTS can be used on anything by default, but to avoid specifying the language in the call it should be configured either on the container or the property. Then, if you want it to be read-efficient, you can also create an index

AndriySvyryd replied:
I think this method should be called HasFullTextSearchLanguage and have a required parameter

AndriySvyryd replied:
And EnableFullTextSearch() is completely unnecessary

maumar replied:
@roji

maumar replied:
language on the path should be optional - there is a bug in current SDK which prevents it - adding compensation for now to fall back to default container language and then to "en-US" if language was not specified for the property using EF layer. Will remove the fallback once Cosmos SDK is updated. For now going with 2 annotation approach proposed by @roji - we can/will revisit the APIs in the future

roji replied:
I've been out of the discussion recently, so I'm probably a bit behind (and I see @AndriySvyryd already approved), so just asking to make sure I understand...

From the discussion offline with the Cosmos people, my understand was that the full-text query functions can be invoked on any text property, without any sort of container definition (basically what @AndriySvyryd wrote here: https://github.com/dotnet/efcore/pull/35868#discussion_r2035677362). I understand there's a discrepancy between this understand and the docs you cited above @maumar, but isn't this something we can clear up quickly with the Cosmos people (e.g. their docs are probably just wrong)?

Because if our understanding from the offline discussion is correct, we indeed shouldn't have EnableFullTextSearch or anything like that - just HasFullTextSearchLanguage (one annotation). We should avoid introducing EnableFullTextSearch and then revisiting to break it later if we can (even if it's marked as experimental).

But if I'm just behind in my understanding feel free to ignore.

roji replied:
OK, discussed offline and the misssing piece for me was that a container FTS definition is required in order to define an index. So it's not just about changing the language - I agree that means that EnableFullTextSearch makes sense.

maumar replied:
spoke with @roji offline, FTS can be used without any setup, but it uses default language and won't use the index, which would be very slow. We want users to be directed into configuring the index, so keeping the current approach for now

roji commented on src/EFCore.Cosmos/Infrastructure/Internal/CosmosModelValidator.cs:
```diff
@@ -570,6 +592,10 @@ protected virtual void ValidateIndexes(
                 }
+                else if (index.IsFullTextIndex() == true)
+                {
+                    // allow these, validation happens during container creation
```
Don't we want to validate here that the FTS index is defined over an FTS property?

maumar replied:
I guess this also is one of those situations that won't suddenly start being supported by cosmos. Per @AndriySvyryd suggestion, I moved most the validation to the container creation step, so that we don't validate & throw for cases that are currently not supported, but might be in the future (e.g. FTS on non-string properties) - if that happens people won't be able to create such databases from scratch, but at least can use them in query. But yeah, case where we define FTS index on non-FTS prop is not one of there, but rather modeling error. So yeah, makes sense to move it here

maumar commented on test/EFCore.Cosmos.FunctionalTests/AddHocFullTextSearchCosmosTest.cs:
```diff
@@ -0,0 +1,415 @@
+        {
+            b.ToContainer("Entities").HasDefaultFullTextLanguage("pl-PL");
+            b.HasPartitionKey(x => x.PartitionKey);
+            b.Property(x => x.Name).EnableFullTextSearch(enabled: true, language: "de-DE");
```
not ideal, but it should be rare. Still seems more discoverable than having two separate methods. Alternatively could add DisableFullTextSearch as a way to un-FTS a property.

maumar commented:
/azp run

AndriySvyryd commented on src/EFCore.Cosmos/Extensions/CosmosPropertyBuilderExtensions.cs:
```diff
@@ -243,4 +243,177 @@ private static CosmosVectorType CreateVectorType(DistanceFunction distanceFuncti
+    [Experimental(EFDiagnostics.CosmosFullTextSearchExperimental)]
+    public static PropertyBuilder EnableFullTextSearch(
+        this PropertyBuilder propertyBuilder,
+        string? language = null)
```
Can add a `bool enabled = true` parameter here to allow disabling FTS
